airbyte-integrations/connectors/source-facebook-marketing/unit_tests/integration/test_ads_insights_action_product_id.py

```diff
@@ -0,0 +1,489 @@
+    }
+    return RequestBuilder.get_insights_endpoint(access_token=ACCESS_TOKEN, account_id=ACCOUNT_ID).with_body(
+        encode_request_body(body)
+    )
+
+
+def _job_status_request(report_run_ids: Union[str, List[str]]) -> RequestBuilder:
+    if isinstance(report_run_ids, str):
+        report_run_ids = [report_run_ids]
+    body = {"batch": [{"method": "GET", "relative_url": f"{report_run_id}/"} for report_run_id in report_run_ids]}
+    return RequestBuilder.get_execute_batch_endpoint(access_token=ACCESS_TOKEN).with_body(encode_request_body(body))
+
+
+def _get_insights_request(job_id: str) -> RequestBuilder:
+    return RequestBuilder.get_insights_download_endpoint(access_token=ACCESS_TOKEN, job_id=job_id).with_limit(100)
+
+
+def _update_api_throttle_limit_response() -> HttpResponse:
+    body = {}
+    headers = {
+        "x-fb-ads-insights-throttle": json.dumps(
+            {"app_id_util_pct": 0, "acc_id_util_pct": 0, "ads_api_access_tier": "standard_access"}
+        ),
+    }
+    return build_response(body=body, status_code=HTTPStatus.OK, headers=headers)
+
+
+def _job_start_response(report_run_id: str) -> HttpResponse:
+    body = {"report_run_id": report_run_id}
+    return build_response(body=body, status_code=HTTPStatus.OK)
+
+
+def _job_status_response(
+    job_ids: Union[str, List[str]], status: Optional[Status] = Status.COMPLETED, account_id: Optional[str] = ACCOUNT_ID
+) -> HttpResponse:
+    if isinstance(job_ids, str):
+        job_ids = [job_ids]
+    body = [
+        {
+            "body": json.dumps(
+                {
+                    "id": job_id, "account_id": account_id, "async_status": status, "async_percent_completion": 100
+                }
+            ),
+        } for job_id in job_ids
+    ]
+    return build_response(body=body, status_code=HTTPStatus.OK)
+
+
+def _insights_response() -> HttpResponseBuilder:
+    return create_response_builder(
+        response_template=find_template(_STREAM_NAME, __file__),
+        records_path=FieldPath("data"),
+        pagination_strategy=FacebookMarketingPaginationStrategy(_get_insights_request(_JOB_ID).with_limit(100).build()),
+    )
+
+
+def _ads_insights_action_product_id_record() -> RecordBuilder:
+    return create_record_builder(
+        response_template=find_template(_STREAM_NAME, __file__),
+        records_path=FieldPath("data"),
+        record_cursor_path=FieldPath(_CURSOR_FIELD),
+    )
+
+
+@freezegun.freeze_time(NOW.isoformat())
+class TestFullRefresh(TestCase):
+
+    @staticmethod
+    def _read(config_: ConfigBuilder, expecting_exception: bool = False) -> EntrypointOutput:
+        return read_output(
+            config_builder=config_,
+            stream_name=_STREAM_NAME,
+            sync_mode=SyncMode.full_refresh,
+            expecting_exception=expecting_exception,
+        )
+
+    @HttpMocker()
+    def test_given_one_page_when_read_then_return_records(self, http_mocker: HttpMocker) -> None:
```

Comment:
This might sound very nit but I think it is very important as not making things explicit doesn't provide the best documentation and hence leads to misunderstanding. I think the test would benefit not to rely too much on constants that are defined in the config. Rewriting the test, it would look like this:

```
    @HttpMocker()
    def test_given_one_page_when_read_then_return_records(self, http_mocker: HttpMocker) -> None:
        end_datetime = _START_DATE + timedelta(hours=23)
        http_mocker.get(get_account_request(_A_CLIENT_SIDE_ACCOUNT_ID).build(), get_account_response(_A_SERVER_SIDE_ACCOUNT_ID))
        http_mocker.get(_update_api_throttle_limit_request(_A_SERVER_SIDE_ACCOUNT_ID).build(), _update_api_throttle_limit_response())
        http_mocker.post(_job_start_request(_A_SERVER_SIDE_ACCOUNT_ID, _START_DATE, end_datetime).build(), _job_start_response(_REPORT_RUN_ID))
        http_mocker.post(_job_status_request(_REPORT_RUN_ID).build(), _job_status_response(_JOB_ID))
        http_mocker.get(
            _get_insights_request(_JOB_ID).build(),
            _insights_response().with_record(_ads_insights_action_product_id_record()).build(),
        )

        output = self._read(config().with_account_ids([_A_CLIENT_SIDE_ACCOUNT_ID]).with_start_date(_START_DATE).with_end_date(end_datetime))
        assert len(output.records) == 1
```

There are a couple of things we can more easily map with the test written like this:
* The start/end datetime provide to the config is used to define boundaries to the job start request
* We only use the account ID from the config at one place and it's when we get the account
* The end_datetime can't be anything here else another behavior would occur (case that you did test elsewhere which is very good)

I would have like to have been explicit like that in Stripe because I could have caught more issues and saved some time during investigations. The specific case I have in mind is that some streams do not consider the replication start. A test that defines `replication_start_date` in the config and not in the endpoint would have expose this very easily

Reply:
Note that I think once we have a test that explicit this, we don't need to be as explicit in the other tests. If the other tests could just highlight the difference that they are testing compared to this "baseline" test, that would be neat as well

Reply:
Totally agree here! Updated `test_given_one_page_when_read_then_return_records` for both `videos` and `ads_insights_action_product_id` streams. Please take a look

Reply:
I like it. Maybe one last thing would be that `client_side_account_id` and `server_side_account_id` should have different value to ensure that if there is a mismatch, the test actually fails

Reply:
Updated, thanks!